modem: backends: uart: improvements

tomi-font opened the pull request:
See details in commit messages.

tomi-font committed:
modem: backends: uart_async: increase default TX timeout

The previous default value made UART TXs of size > 1145
time out at a baud rate of 115200.

It should not be so small that it can cause undesirable timeouts.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

tomi-font committed:
samples: net: cellular_modem: increase the post-restart sleep time

A higher delay helps in getting the first post-restart echo
packet to be successful.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

tomi-font committed:
modem: backends: uart: register their own log modules

Instead of both using the same modem_backend_uart, for better visibility.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

tomi-font committed:
modem: cmux: improve frame debug logging

- Add the frame data length to the frame header log.
- Log the dropped frames as well.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

tomi-font committed:
modem: backends: uart_async: small improvements

- Set all the state bits before calling uart_rx_enable() because
  interrupts may happen already before it returns.
- Fix the byte count in the "Receive buffer overrun" log.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

tomi-font committed:
modem: backends: uart_isr: improve the reception of bytes

Add a configurable delay between when a byte is received and
MODEM_PIPE_EVENT_RECEIVE_READY is sent.

This fixes data reception at baud rates above 460800, and
most likely also reduces the workload at any baud rate
when receiving bytes by not going through the work item
and callbacks for every single byte.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

juser19201 commented:
I had a discussion yesterday with @bjarki-trackunit about getting frequent and random `modem_cmux: Frame FCS error` and `modem_cmux: Dropped frame`, this PR seems to fix my issue. I get the same warnings very infrequently, but with the `uart_backend_receive_buf` being cofigurable by a KConfig, making that bigger and then the changes from this PR fixed my issue.

rerickson1 commented on subsys/modem/modem_cmux.c:
```diff
@@ -842,7 +842,7 @@ static void modem_cmux_receive_handler(struct k_work *item)
 	struct k_work_delayable *dwork = k_work_delayable_from_work(item);
 	struct modem_cmux *cmux = CONTAINER_OF(dwork, struct modem_cmux, receive_work);
-	uint8_t buf[16];
+	uint8_t buf[32];
```
Should there be a define for this value? Might make it more clear why this size was chosen.

tomi-font replied:
Yeah, could be. I wanted to increase the size of this buffer to further reduce the number of calls to `modem_pipe_receive()` (which is what the ISR UART improvement in this PR is about).
I took 32 as I saw that's what `modem_chat` uses for its `work_buf`, and to keep it still relatively small as here the buffer lies on the stack.

The buffer size could even be made configurable in my opinion. What do you think? @bjarki-trackunit

bjarki-trackunit replied:
This buffer is a bit ugly, the reason I added it to the stack here was to avoid adding yet another buffer one would have to allocate and pass when initializing the cmux instance. The best option is probably to just add the buffer, naming it cache_buffer or something, but Kconfig is definately the simplest option :)

tomi-font replied:
I was rather thinking of adding the buffer to the `modem_cmux` struct. Any objection?

bjarki-trackunit replied:
oh sure, that is way better, that and the Kconfig is perfect :)

rerickson1 reviewed (dismissed):
Tested on the Pinnacle 100 modem (ASYNC UART). Works as expected.

rerickson1 commented:
> Curiously, `CONFIG_UART_INTERRUPT_DRIVEN=y` is not working on the Pinnacle 100 (nrf52840 + HL7800).

Tried this on main and the issue is still there. So this PR is not the cause.

tomi-font commented:
> > Curiously, `CONFIG_UART_INTERRUPT_DRIVEN=y` is not working on the Pinnacle 100 (nrf52840 + HL7800).
> 
> Tried this on main and the issue is still there. So this PR is not the cause.

Interesting. Has it worked in the past?

jukkar commented on subsys/modem/backends/Kconfig:
```diff
@@ -22,11 +22,28 @@ config MODEM_BACKEND_UART_ASYNC
+if MODEM_BACKEND_UART_ISR
+
+config MODEM_BACKEND_UART_ISR_RECEIVE_IDLE_TIMEOUT_MS
+	int "Modem ISR UART backend delay between first byte received and RECEIVE_READY pipe event sending"
+	default 20
+	help
+	  This defines the delay, in milliseconds, that the backend waits when receiving a byte before notifying that data is available.
+	  This is to avoid having the receiver call modem_pipe_receive() too often (e.g. for every byte).
+	  It temporarily disables the UART RX IRQ when swapping buffers which can cause byte loss at higher baud rates.
+	  Regardless of this value, when a receive buffer gets full the event to notify that data is available will be sent.
+	  A good value for this is is slightly less than (~90%) the time it takes to fill up one such buffer.
+	  It can be calculated as follows:
+	  (<UART receive_buf_size> / 2) / (<UART baud rate> / <UART bits per byte>) * <ms per sec>
+	  By default (for the modem_cellular driver): (512 / 2) / (115200 / 10) * 1000 = 22,222 => 20
```
Very long lines here, I wonder if we could make the option title a bit shorter (to make it more user friendly when using `menuconfig` command).
Also the help text lines are quite log, perhaps they could be split to smaller pieces.

jukkar commented on subsys/modem/backends/modem_backend_uart.c:
```diff
@@ -9,15 +9,12 @@
-	struct modem_backend_uart *backend =
-		CONTAINER_OF(item, struct modem_backend_uart, receive_ready_work);
+	struct modem_backend_uart *backend = CONTAINER_OF(
+		(struct k_work_delayable *)item, struct modem_backend_uart, receive_ready_work);
```
There is the `k_work_delayable_from_work(struct k_work *work);` that you should use instead of direct cast.

jukkar commented on subsys/modem/modem_cmux.c:
```diff
@@ -652,6 +656,19 @@ static void modem_cmux_on_frame(struct modem_cmux *cmux)
+	LOG_WRN("Dropped frame");
+	cmux->receive_state = MODEM_CMUX_RECEIVE_STATE_SOF;
+
+#if CONFIG_MODEM_CMUX_LOG_LEVEL_DBG
```
Better to use `#if defined(CONFIG_MODEM_CMUX_LOG_LEVEL_DBG)` as that will not give error if the define does not exists.

rerickson1 commented:
> > > Curiously, `CONFIG_UART_INTERRUPT_DRIVEN=y` is not working on the Pinnacle 100 (nrf52840 + HL7800).
> > 
> > 
> > Tried this on main and the issue is still there. So this PR is not the cause.
> 
> Interesting. Has it worked in the past?

At this point I can't remember ;). I'll raise a bug.

rerickson1 commented:
bug: https://github.com/zephyrproject-rtos/zephyr/issues/70498

tomi-font committed:
modem: cmux: make work buffer size configurable

A Kconfig item is added and the buffer
is embedded into the modem_cmux struct.
The default value is increased from 16 to 64 bytes in
an effort to reduce the number of modem_pipe_receive() calls.

CONFIG_MODEM_CHAT_LOG_BUFFER is renamed
to CONFIG_MODEM_CHAT_LOG_BUFFER_SIZE for consistency.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>

rerickson1 commented on subsys/modem/Kconfig:
```diff
@@ -14,8 +14,8 @@ config MODEM_CHAT
 
-config MODEM_CHAT_LOG_BUFFER
-	int "Modem chat log buffer size"
+config MODEM_CHAT_LOG_BUFFER_SIZE
+	int "Modem chat log buffer size in bytes"
```
This affects users... Do we need this mentioned in release notes some where?

rlubos replied:
I think migration guide would be the place:
https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/releases/migration-guide-3.7.rst

bjarki-trackunit replied:
Good catch! We need to add this change to the migration guide :)

tomi-font commented:
Does anything need to be mentioned in the release notes also? (e.g. the new Kconfig option)

jukkar commented:
> Does anything need to be mentioned in the release notes also? (e.g. the new Kconfig option)

I don't think so, the migration guide addition should be enough here.

tomi-font commented:
ping @bjarki-trackunit
